File: android/app/src/main/java/sh/mo/TranscriptState.kt
package sh.mo

/**
 * Transcript-state reducer, app side (pure Kotlin, no Android imports — the
 * JVM test suite links this directly).
 *
 * This is where the transcript-fidelity defect lived. The core's changes
 * arrive as (kind, index, rowJson); the old repository code applied them
 * straight into `SessionUiState.rows` as plain strings, with two defects:
 *
 * 1. ROW_UPDATED at an index past the end of the local list was DROPPED
 *    (the stream keeps a contiguous list; any dropped update desynchronises
 *    every later index — rows silently collapse during a turn);
 * 2. RESET wiped the rows and nothing rebuilt them: the desktop renders the
 *    resume payload's own `messages`, but this app's transcript is ONLY what
 *    the change stream delivered, so a Reset left the screen empty until a
 *    fresh turn refilled it.
 *
 * The fix, in one place: a RESET carries the last-known full snapshot (the
 * core re-delivers the rebuilt transcript through the change stream after
 * the reset — see `replayInto`), a ROW_UPDATED past the end is treated as an
 * APPEND, and a gap is padded with empty rows rather than dropping the
 * update. Rows are keyed by their transcript index, so the reducer is
 * total: any sequence of change kinds yields a list whose size covers every
 * index ever referenced.
 */
data class SessionUiState(
    val key: String,
    val title: String = "",
    val model: String = "",
    val rows: List<String> = emptyList(),
    /** True while the session's turn is streaming (SessionSummary.running). */
    val running: Boolean = false,
    /**
     * T16b: which Hermes profile owns the chat (SessionSummary.profile_name),
     * carried on the tab model for T16c's drawer — NOT rendered this layer.
     * Empty when unknown — never a guess.
     */
    val profile: String = "",
) {
    fun withRows(rows: List<String>) = copy(rows = rows)
}

/**
 * One row of the T11 session picker (the phone sheet over `session.list`):
 * `RemoteSessionDto` flattened into a pure value the sheet renders. Pure
 * Kotlin so the JVM suite can pin the title/preview fallbacks.
 */
data class RemoteSessionRow(
    val id: String,
    val title: String,
    val preview: String,
    val messageCount: Long,
) {
    /** Titlebar copy: the session's title, or "New session" when unnamed. */
    val displayTitle: String get() = title.ifBlank { "New session" }

    /** The one-line preview, blank when the session has no text yet. */
    val displayPreview: String get() = preview.trim()
}

/**
 * Apply one change-stream event to the per-key map.
 *
 * FIX7 (review 5214081721, finding 1): the map may only GROW for a key the
 * repository considers open-or-restoring ([knownKeys] = the live `TabSet.keys`
 * union the keys the launch restore plan pre-registered). The pre-fix code
 * minted a [SessionUiState] for EVERY arriving key — contradicting this
 * file's own contract and letting a late transcript change for a key closed
 * via `closeTab` resurrect the phantom entry. Unknown and closed keys are
 * now DROPPED.
 *
 * The resume race stays covered at the adapter: the repository puts the
 * restore keys into `knownKeys` BEFORE the sink is allowed to deliver
 * (register, then consume), so a resume's Reset+rows still land even when
 * they race tab registration. `currentKey` is never set by an arriving
 * change — that invariant is untouched.
 */
fun applySessionChange(
    sessions: Map<String, SessionUiState>,
    key: String,
    kind: String,
    index: Long,
    rowJson: String,
    knownKeys: Set<String>,
): Map<String, SessionUiState> {
    // The header branch (finding 4) sits behind the same gate: a header for
    // an unknown/closed key must not mint an empty entry either.
    if (key !in knownKeys) return sessions
    val current = sessions[key] ?: SessionUiState(key = key)
    if (kind == "headerUpdated") return sessions + (key to current)
    return sessions + (key to current.copy(rows = applyTranscriptChange(current.rows, kind, index, rowJson)))
}

/** Register a tab's key without clobbering rows the stream already delivered. */
fun ensureSession(sessions: Map<String, SessionUiState>, key: String): Map<String, SessionUiState> =
    if (key in sessions) sessions else sessions + (key to SessionUiState(key = key))

/**
 * Apply one transcript change to `rows`. `rowJson` is the core's row JSON
 * (empty for RESET). Returns the new list — a pure function, trivially
 * testable.
 */
fun applyTranscriptChange(rows: List<String>, kind: String, index: Long, rowJson: String): List<String> {
    return when (kind) {
        "rowAppended" -> {
            val out = rows.toMutableList()
            // A replayed append for an index we already hold must not
            // duplicate the row (the reconnect replay can repeat appends).
            while (out.size <= index.toInt()) out.add("")
            out[index.toInt()] = rowJson
            out
        }
        "rowUpdated" -> {
            val i = index.toInt()
            val out = rows.toMutableList()
            if (i >= out.size) {
                // THE DEFECT: the old code dropped this update, desynchronising
                // every later index. Pad the gap, then set — the row list stays
                // contiguous and no update is ever lost.
                while (out.size <= i) out.add("")
            }
            out[i] = rowJson
            out
        }
        "reset" -> {
            // T7c: the Reset clears the transcript and nothing app-side
            // replays a snapshot — the rebuilt rows arrive right after as
            // RowAppended changes from the core (the stream is the single
            // source of rows).
            emptyList()
        }
        else -> rows // headerUpdated: no row change
    }
}

/**
 * Incremental parse of the raw row JSON into [ChatRow]s.
 *
 * Two things this exists for.
 *
 * Identity: a row's id is its TRANSCRIPT index, not its position after the
 * empty-row filter. `applyTranscriptChange` pads gaps with empty strings, so a
 * filtered position shifts for every row after a gap the moment that gap is
 * filled — and the transcript list keys on the id, so every later row looked
 * like a brand new item: scroll position jumped and per-row state (expanded
 * tool cards, thinking disclosures) reset mid-turn.
 *
 * Cost: a streaming turn rewrites ONE row's JSON per delta while recomposition
 * happens for unrelated reasons too (the elapsed-time tick, IME insets, focus).
 * Re-parsing the whole transcript each time is O(rows) of `JSONObject` per
 * token; reusing the previous parse for every row whose JSON is unchanged
 * makes it O(changed).
 */
class TranscriptRows {
    private var lastRaw: List<String> = emptyList()
    private var lastParsed: List<ChatRow?> = emptyList()
    private var lastVisible: List<ChatRow> = emptyList()

    fun of(raw: List<String>): List<ChatRow> {
        // Reference equality, not structural: the same instance means nothing
        // moved, and a structural compare would cost the per-row string
        // comparison the incremental path below already does.
        if (raw === lastRaw) return lastVisible
        val parsed = ArrayList<ChatRow?>(raw.size)
        for (index in raw.indices) {
            val json = raw[index]
            val reusable = index < lastRaw.size && lastRaw[index] == json
            parsed.add(
                when {
                    reusable -> lastParsed[index]
                    json.isEmpty() -> null
                    else -> parseChatRow(index, json)
                },
            )
        }
        lastRaw = raw
        lastParsed = parsed
        lastVisible = parsed.filterNotNull()
        return lastVisible
    }
}

File: ios/Sources/HermoLogic/TranscriptState.swift
import Foundation

/// One row of the session picker sheet (the phone's view of `session.list`):
/// `RemoteSessionDto` flattened into a pure value the sheet renders. Pure so the
/// logic suite can pin the title and preview fallbacks without the gateway.
public struct RemoteSessionRow: Equatable, Sendable, Identifiable {
    public let id: String
    public let title: String
    public let preview: String
    public let messageCount: Int64

    public init(id: String, title: String, preview: String, messageCount: Int64) {
        self.id = id
        self.title = title
        self.preview = preview
        self.messageCount = messageCount
    }

    /// Titlebar copy: the session's title, or "New session" when unnamed.
    public var displayTitle: String {
        title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "New session" : title
    }

    /// The one-line preview, blank when the session has no text yet.
    public var displayPreview: String {
        preview.trimmingCharacters(in: .whitespacesAndNewlines)
    }
}

/// Applies one change-stream event to the per-key session map. A key outside
/// `knownKeys` (not an open-or-restoring tab) is dropped rather than minted,
/// including for `headerUpdated`, so a late change for a key already closed
/// via `closeTab` cannot resurrect a phantom entry.
public func applySessionChange(
    sessions: [String: SessionUiState],
    key: String,
    kind: String,
    index: Int64,
    rowJson: String,
    knownKeys: Set<String>,
    title: String = ""
) -> [String: SessionUiState] {
    guard knownKeys.contains(key) else { return sessions }
    var current = sessions[key] ?? SessionUiState(key: key)
    var updated = sessions
    if kind == "headerUpdated" {
        // An empty title means "no rename on this event" and must never blank a known title.
        if !title.isEmpty { current.title = title }
        updated[key] = current
        return updated
    }
    current.rows = applyTranscriptChange(current.rows, kind: kind, index: index, rowJson: rowJson)
    updated[key] = current
    return updated
}

/// Registers a tab's key without disturbing rows the change stream already delivered for it.
public func ensureSession(_ sessions: [String: SessionUiState], _ key: String) -> [String: SessionUiState] {
    if sessions[key] != nil { return sessions }
    var updated = sessions
    updated[key] = SessionUiState(key: key)
    return updated
}

/// Applies one transcript change to `rows`, keyed by transcript index rather
/// than list position. `rowUpdated` past the end of the list pads with empty
/// rows instead of dropping the update, which used to desynchronise every
/// index after the gap. `reset` clears the list; the core re-delivers the
/// rebuilt transcript right after as one `rowAppended` per row.
public func applyTranscriptChange(_ rows: [String], kind: String, index: Int64, rowJson: String) -> [String] {
    switch kind {
    case "rowAppended", "rowUpdated":
        // A negative index can only come from a malformed change; padding to it
        // is impossible and subscripting with it would trap.
        guard index >= 0, index <= Int64(Int.max) else { return rows }
        let i = Int(index)
        var out = rows
        while out.count <= i { out.append("") }
        out[i] = rowJson
        return out
    case "reset":
        return []
    default:
        return rows
    }
}

/// Incremental transcript parser, keyed by transcript index. `applyTranscriptChange`
/// pads gaps with empty strings, so identity has to survive on the index, not
/// on filtered list position, or every row after a gap would look like a new
/// item the moment the gap fills in.
///
/// A streaming turn rewrites one row's JSON per delta while unrelated
/// recomposition (elapsed-time tick, IME insets, focus) fires far more often,
/// so a row whose JSON string is unchanged since the last call reuses its
/// previously parsed value instead of re-parsing. Swift arrays carry no cheap
/// reference identity the way a Kotlin `List` does, so the whole-list
/// short-circuit the Kotlin version takes before this loop is dropped; the
/// per-index string comparison below is what actually saves the work, and it
/// runs unconditionally here.
public final class TranscriptRows {
    private var lastRaw: [String] = []
    private var lastParsed: [ChatRow?] = []

    public init() {}

    public func of(_ raw: [String]) -> [ChatRow] {
        var parsed: [ChatRow?] = []
        parsed.reserveCapacity(raw.count)
        for index in raw.indices {
            let json = raw[index]
            if index < lastRaw.count, lastRaw[index] == json {
                parsed.append(lastParsed[index])
            } else if json.isEmpty {
                parsed.append(nil)
            } else {
                parsed.append(parseChatRow(index, json))
            }
        }
        lastRaw = raw
        lastParsed = parsed
        return parsed.compactMap { $0 }
    }
}
